File: modules/tif-haptics/android/src/main/java/expo/modules/tifhaptics/TifHapticsModule.kt
package expo.modules.tifhaptics

import android.content.Context
import com.tifapp.tifnative.AndroidHapticsPlayer
import com.tifapp.tifnative.ExpoObject
import com.tifapp.tifnative.isHapticsSupportedOnDevice
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class TifHapticsModule : Module() {
  private val context: Context
    get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()

  override fun definition() = ModuleDefinition {
    val haptics = AndroidHapticsPlayer(context)
    Name("TifHaptics")

    AsyncFunction("play") Coroutine { pattern: ExpoObject ->
      haptics.playPattern(pattern)
    }

    AsyncFunction("apply") Coroutine { settings: ExpoObject ->
      haptics.apply(settings)
    }

    Function("deviceSupport") {
      mapOf(
        "isFeedbackSupportedOnDevice" to context.isHapticsSupportedOnDevice,
        "isAudioSupportedOnDevice" to true
      )
    }
  }
}

File: modules/tif-haptics/ios/TifHapticsModule.swift
import ExpoModulesCore
import CoreHaptics
import TiFNative

public class TifHapticsModule: Module {
  private let engine = TiFHapticsEngine()

  public func definition() -> ModuleDefinition {
    Name("TifHaptics")
    
    Function("deviceSupport") {
      let capabilities = CHHapticEngine.capabilitiesForHardware()
      return [
        "isFeedbackSupportedOnDevice": capabilities.supportsHaptics,
        "isAudioSupportedOnDevice": capabilities.supportsAudio
      ]
    }

    AsyncFunction("apply") { (settings: ExpoObject) async throws -> Void in
      let settings = try HapticsSettings(expoObject: settings)
      try await self.engine.apply(settings: settings)
    }

    AsyncFunction("play") { (event: ExpoObject) async throws -> Void in
      try await self.engine.play(expoPattern: event)
    }
  }
}
